The BLT provided "openmp" pseudotarget doesn't (always) pass the correct openMP fortran flags

The "openmp" pseudotarget assumes that the OpenMP Fortran Flags are the same as C/C++ compiler flags.

We register the "openmp" pseudotarget as follows:
```
#################################################
# OpenMP
# (OpenMP support is provided by the compiler)
#################################################
message(STATUS "OpenMP Support is ${ENABLE_OPENMP}")
if(ENABLE_OPENMP)
    find_package(OpenMP REQUIRED)
    message(STATUS "OpenMP CXX Flags: ${OpenMP_CXX_FLAGS}")
    
    # register openmp with blt
    blt_register_library(NAME openmp
                         COMPILE_FLAGS ${OpenMP_CXX_FLAGS} 
                         LINK_FLAGS    ${OpenMP_CXX_FLAGS} )
endif()
```

This works for the most part, as long as:
* The C++ language is enabled
* The C++ and Fortran OpenMP flags match 

The latter may not always be the case. For example, for a project that uses Clang for C/C++ in conjunction with GNU Fortran, CMake will compute the following flags:

```
//C++ compiler flags for OpenMP parallization
OpenMP_CXX_FLAGS:STRING=-fopenmp=libomp

//C compiler flags for OpenMP parallization
OpenMP_C_FLAGS:STRING=-fopenmp=libomp

//Fortran compiler flags for OpenMP parallization
OpenMP_Fortran_FLAGS:STRING=-fopenmp
```

But, when building an executable like this:
```
blt_add_executable(
    NAME foo
    SOURCES foo.F
    DEPENDS_ON  openmp
    )
```

This would pass `OpenMP_CXX_FLAGS`, leading to the following compiler error:
```
gfortran: error: unrecognized command line option '-fopenmp=libomp'
```

Perhaps, BLT should register two pseudo libraries/targets for C/C++ and Fortran respectively, like this:
```
if(ENABLE_OPENMP)
    find_package(OpenMP REQUIRED)
    message(STATUS "OpenMP CXX Flags: ${OpenMP_CXX_FLAGS}")
    
    # register openmp with blt
    blt_register_library(NAME openmp-cxx
                         COMPILE_FLAGS ${OpenMP_CXX_FLAGS} 
                         LINK_FLAGS    ${OpenMP_CXX_FLAGS} )

   if ( ENABLE_FORTRAN)
      blt_register_library(NAME openmp-fortran
                                       COMPILE_FLAGS ${OpenMP_Fortran_FLAGS}
                                       LINK_FLAGS ${OpenMP_Fortran_FLAGS} )
  endif()

endif()
```

This would at least make it more explicit to the application using this about how things are compiled/linked. Or alternatively, come up with a way to make it easier for an application to set source file properties for OpenMP. The latter may allow the most flexibility and be less clunky.

Is this the symptom:
```
-- Configuring done
CMake Error at CMakeLists.txt:82 (target_link_libraries):
  Error evaluating generator expression:
    $<COMPILE_LANGUAGE:Fortran>
  $<COMPILE_LANGUAGE:...> Unknown language.
-- Generating done
```

```
I/usr/WS2/larsen30/ascent_build/ascent/uberenv_libs/spack/opt/spack/linux-rhel7-x86_64/gcc-4.9.3/zlib-1.2.11-3r4cfkmx3wwfqeof4bc244yduu2mz4ur/include;\$<\$<NOT:\$<COMPILE_LANGUAGE:Fortran>>:-fopenmp>;\$<\$<COMPILE_LANGUAGE:Fortran>:-fopenmp>"
```

--- a/cmake/BLTMacros.cmake
+++ b/cmake/BLTMacros.cmake
@@ -561,6 +561,12 @@ macro(blt_add_executable)
                      DEPENDS_ON ${arg_DEPENDS_ON} 
                      OBJECT     FALSE)
 
+    # fix the openmp flags for fortran if needed
+    # NOTE: this needs to be called after blt_setup_target()
+    if (_lang STREQUAL Fortran)
+       blt_fix_fortran_openmp_flags( ${arg_NAME} )
+    endif()
+
     if ( arg_INCLUDES )
         target_include_directories(${arg_NAME} PUBLIC ${arg_INCLUDES})
     endif()

--- a/cmake/BLTPrivateMacros.cmake
+++ b/cmake/BLTPrivateMacros.cmake
@@ -1,6 +1,6 @@
 # Copyright (c) 2017-2019, Lawrence Livermore National Security, LLC and
 # other BLT Project Developers. See the top-level COPYRIGHT file for details
-# 
+#
 # SPDX-License-Identifier: (BSD-3-Clause)
 
 include(CMakeParseArguments)
@@ -83,6 +83,34 @@ function(blt_error_if_target_exists target_name error_msg)
     endif()
 endfunction()
 
+##-----------------------------------------------------------------------------
+## blt_fix_fortran_openmp_flags()
+##
+## Fixes the openmp flags for a Fortran target if they are different from the
+## corresponding C/C++ openmp flags.
+##-----------------------------------------------------------------------------
+function(blt_fix_fortran_openmp_flags target_name)
+
+
+    if (ENABLE_FORTRAN AND ENABLE_OPENMP AND BLT_OPENMP_FLAGS_DIFFER)
+
+        get_target_property(target_link_flags ${target_name} LINK_FLAGS)
+        if ( target_link_flags )
+
+            message(STATUS "Fixing OpenMP Flags for target[${target_name}]")
+
+            string( REPLACE "${OpenMP_CXX_FLAGS}" "${OpenMP_Fortran_FLAGS}"
+                    correct_link_flags
+                    "${target_link_flags}"
+                    )
+
+            set_target_properties( ${target_name} PROPERTIES LINK_FLAGS
+                                   "${correct_link_flags}" )
+        endif()
+
+    endif()
+
+endfunction()
 
 ##-----------------------------------------------------------------------------
 ## blt_find_executable(NAME         <name of program to find>
@@ -213,8 +241,8 @@ endmacro(blt_inherit_target_info)
 
 
 ##------------------------------------------------------------------------------
-## blt_setup_target( NAME       [name] 
-##                   DEPENDS_ON [dep1 ...] 
+## blt_setup_target( NAME       [name]
+##                   DEPENDS_ON [dep1 ...]
 ##                   OBJECT     [TRUE | FALSE])
 ##------------------------------------------------------------------------------
 macro(blt_setup_target)
@@ -351,7 +379,7 @@ macro(blt_setup_cuda_target)
     endif()
 
     if (${_depends_on_cuda_runtime} OR ${_depends_on_cuda})
-        if (CUDA_LINK_WITH_NVCC) 
+        if (CUDA_LINK_WITH_NVCC)
             set_target_properties( ${arg_NAME} PROPERTIES LINKER_LANGUAGE CUDA)
         endif()
     endif()
@@ -617,7 +645,7 @@ macro(blt_filter_list)
     set(multiValueArgs )
 
     # Parse arguments
-    cmake_parse_arguments(arg "${options}" "${singleValueArgs}" 
+    cmake_parse_arguments(arg "${options}" "${singleValueArgs}"
                             "${multiValueArgs}" ${ARGN} )
 
     # Check arguments
@@ -678,7 +706,7 @@ macro(blt_clean_target)
     set(multiValueArgs )
 
     # Parse arguments
-    cmake_parse_arguments(arg "${options}" "${singleValueArgs}" 
+    cmake_parse_arguments(arg "${options}" "${singleValueArgs}"
                             "${multiValueArgs}" ${ARGN} )
 
     # Properties to remove duplicates from

--- a/cmake/thirdparty/SetupOpenMP.cmake
+++ b/cmake/thirdparty/SetupOpenMP.cmake
@@ -10,7 +10,15 @@
 
 find_package(OpenMP REQUIRED)
 
-# avoid generator expressions if possible, as generator expressions can be 
+# check if the openmp flags used for C/C++ are different from the openmp flags
+# used by the Fortran compiler
+set(BLT_OPENMP_FLAGS_DIFFER FALSE CACHE BOOL "")
+if (ENABLE_FORTRAN)
+  string(COMPARE NOTEQUAL "${OpenMP_CXX_FLAGS}" "${OpenMP_Fortran_FLAGS}"
+         BLT_OPENMP_FLAGS_DIFFER )
+endif()
+
+# avoid generator expressions if possible, as generator expressions can be
 # passed as flags to downstream projects that might not be using the same
 # languages. See https://github.com/LLNL/blt/issues/205
 if (ENABLE_FORTRAN AND NOT OpenMP_CXX_FLAGS STREQUAL OpenMP_Fortran_FLAGS)

--- a/cmake/thirdparty/SetupOpenMP.cmake
+++ b/cmake/thirdparty/SetupOpenMP.cmake
@@ -21,16 +21,10 @@ endif()
 # avoid generator expressions if possible, as generator expressions can be
 # passed as flags to downstream projects that might not be using the same
 # languages. See https://github.com/LLNL/blt/issues/205
-if (ENABLE_FORTRAN AND NOT OpenMP_CXX_FLAGS STREQUAL OpenMP_Fortran_FLAGS)
-   set(ESCAPE_FORTRAN ON)
-else()
-   set(ESCAPE_FORTRAN OFF)
-endif()
-
 set(_compile_flags ${OpenMP_CXX_FLAGS})
 set(_link_flags  ${OpenMP_CXX_FLAGS})
 
-if(NOT COMPILER_FAMILY_IS_MSVC AND ENABLE_CUDA AND ESCAPE_FORTRAN)
+if(NOT COMPILER_FAMILY_IS_MSVC AND ENABLE_CUDA AND BLT_OPENMP_FLAGS_DIFFER)
     set(_compile_flags
         $<$<AND:$<NOT:$<COMPILE_LANGUAGE:CUDA>>,$<NOT:$<COMPILE_LANGUAGE:Fortran>>>:${OpenMP_CXX_FLAGS}> 
         $<$<COMPILE_LANGUAGE:CUDA>:-Xcompiler=${OpenMP_CXX_FLAGS}>
@@ -39,7 +33,7 @@ elseif(NOT COMPILER_FAMILY_IS_MSVC AND ENABLE_CUDA)
     set(_compile_flags
         $<$<NOT:$<COMPILE_LANGUAGE:CUDA>>:${OpenMP_CXX_FLAGS}> 
         $<$<COMPILE_LANGUAGE:CUDA>:-Xcompiler=${OpenMP_CXX_FLAGS}>)
-elseif(NOT COMPILER_FAMILY_IS_MSVC AND ESCAPE_FORTRAN)
+elseif(NOT COMPILER_FAMILY_IS_MSVC AND BLT_OPENMP_FLAGS_DIFFER)
     set(_compile_flags
         $<$<NOT:$<COMPILE_LANGUAGE:Fortran>>:${OpenMP_CXX_FLAGS}>
         $<$<COMPILE_LANGUAGE:Fortran>:${OpenMP_Fortran_FLAGS}>)

